File: android/app/src/main/java/com/ccoan/ny/ui/screens/projects/ProjectsViewModel.kt
package com.ccoan.ny.ui.screens.projects

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ccoan.ny.data.models.*
import com.ccoan.ny.data.repository.ProjectRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

enum class ViewMode {
    KANBAN, LIST, CALENDAR
}

@HiltViewModel
class ProjectsViewModel @Inject constructor(
    private val repository: ProjectRepository
) : ViewModel() {

    val projects = repository.projects
    val departments = repository.departments

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _viewMode = MutableStateFlow(ViewMode.KANBAN)
    val viewMode: StateFlow<ViewMode> = _viewMode.asStateFlow()

    private val _selectedDepartmentId = MutableStateFlow<String?>(null)
    val selectedDepartmentId: StateFlow<String?> = _selectedDepartmentId.asStateFlow()

    private val _selectedPriority = MutableStateFlow<Priority?>(null)
    val selectedPriority: StateFlow<Priority?> = _selectedPriority.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    val filteredProjects: StateFlow<List<Project>> = combine(
        projects,
        _selectedDepartmentId,
        _selectedPriority,
        _searchQuery
    ) { projectList, deptId, priority, query ->
        projectList.filter { project ->
            // Department filter
            if (deptId != null && project.departmentId != deptId) return@filter false
            // Priority filter
            if (priority != null && project.priority != priority) return@filter false
            // Search filter
            if (query.isNotEmpty()) {
                val q = query.lowercase()
                if (!project.title.lowercase().contains(q) &&
                    !project.description.lowercase().contains(q)) {
                    return@filter false
                }
            }
            // Hide archived
            project.status != ProjectStatus.ARCHIVED
        }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun projectsForStatus(status: ProjectStatus): List<Project> {
        return filteredProjects.value
            .filter { it.status == status }
            .sortedBy { it.order }
    }

    fun loadData() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            val projectsResult = repository.loadProjects()
            val deptResult = repository.loadDepartments()

            projectsResult.onFailure { _errorMessage.value = it.message }
            deptResult.onFailure { _errorMessage.value = it.message }

            _isLoading.value = false
        }
    }

    fun setViewMode(mode: ViewMode) {
        _viewMode.value = mode
    }

    fun setDepartmentFilter(deptId: String?) {
        _selectedDepartmentId.value = deptId
    }

    fun setPriorityFilter(priority: Priority?) {
        _selectedPriority.value = priority
    }

    fun setSearchQuery(query: String) {
        _searchQuery.value = query
    }

    fun moveProject(projectId: String, newStatus: ProjectStatus) {
        viewModelScope.launch {
            repository.reorderProject(projectId, newStatus, 0)
        }
    }

    fun deleteProject(id: String) {
        viewModelScope.launch {
            repository.deleteProject(id)
        }
    }

    fun addComment(projectId: String, text: String) {
        viewModelScope.launch {
            repository.addComment(projectId, text)
        }
    }

    fun createProject(request: CreateProjectRequest, onSuccess: () -> Unit) {
        viewModelScope.launch {
            _isLoading.value = true
            val result = repository.createProject(request)
            result.onSuccess { onSuccess() }
            result.onFailure { _errorMessage.value = it.message }
            _isLoading.value = false
        }
    }

    fun clearError() {
        _errorMessage.value = null
    }
}

File: ios/CCOAN/Features/Projects/ViewModels/ProjectsViewModel.swift
import Foundation
import SwiftUI

@MainActor
class ProjectsViewModel: ObservableObject {
    @Published var projects: [Project] = []
    @Published var departments: [Department] = []
    @Published var users: [User] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    // Filters
    @Published var selectedDepartmentId: String? = nil
    @Published var selectedPriority: Priority? = nil
    @Published var showMyTasksOnly = false
    @Published var searchQuery = ""
    @Published var viewMode: ViewMode = .kanban

    enum ViewMode: String, CaseIterable {
        case kanban = "Board"
        case list = "List"
        case calendar = "Calendar"
    }

    private let apiClient = APIClient.shared

    var filteredProjects: [Project] {
        projects.filter { project in
            // Department filter
            if let deptId = selectedDepartmentId, project.departmentId != deptId {
                return false
            }

            // Priority filter
            if let priority = selectedPriority, project.priority != priority {
                return false
            }

            // Search filter
            if !searchQuery.isEmpty {
                let query = searchQuery.lowercased()
                if !project.title.lowercased().contains(query) &&
                   !project.description.lowercased().contains(query) {
                    return false
                }
            }

            // Don't show archived in normal view
            if project.status == .archived {
                return false
            }

            return true
        }
    }

    func projectsForStatus(_ status: ProjectStatus) -> [Project] {
        filteredProjects
            .filter { $0.status == status }
            .sorted { $0.order < $1.order }
    }

    var taskStats: (total: Int, completed: Int, overdue: Int) {
        let total = filteredProjects.count
        let completed = filteredProjects.filter { $0.status == .done }.count
        let overdue = filteredProjects.filter { project in
            guard let dueDate = project.dueDate else { return false }
            let formatter = DateFormatter()
            formatter.dateFormat = "yyyy-MM-dd"
            guard let date = formatter.date(from: dueDate) else { return false }
            return date < Date() && project.status != .done
        }.count
        return (total, completed, overdue)
    }

    func loadData() async {
        isLoading = true
        errorMessage = nil

        do {
            async let projectsResponse: APIResponse<[Project]> = apiClient.get(endpoint: .projects)
            async let departmentsResponse: APIResponse<[Department]> = apiClient.get(endpoint: .departments)

            let (projRes, deptRes) = try await (projectsResponse, departmentsResponse)

            if let projectsData = projRes.data {
                self.projects = projectsData
            }
            if let departmentsData = deptRes.data {
                self.departments = departmentsData
            }
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func loadUsers() async {
        do {
            let response: APIResponse<[User]> = try await apiClient.get(endpoint: .users)
            if let usersData = response.data {
                self.users = usersData
            }
        } catch {
            // Users might not be accessible for non-superusers
        }
    }

    func createProject(_ request: CreateProjectRequest) async -> Bool {
        do {
            let response: APIResponse<CreateProjectResponse> = try await apiClient.post(
                endpoint: .projects,
                body: request
            )
            if response.success {
                await loadData()
                return true
            } else {
                errorMessage = response.error ?? "Failed to create task"
            }
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }
        return false
    }

    func updateProject(_ request: UpdateProjectRequest) async -> Bool {
        do {
            let response: APIResponse<EmptyResponse> = try await apiClient.put(
                endpoint: .projects,
                body: request
            )
            if response.success {
                await loadData()
                return true
            } else {
                errorMessage = response.error ?? "Failed to update task"
            }
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }
        return false
    }

    func moveProject(projectId: String, to status: ProjectStatus, order: Int = 0) async {
        do {
            let request = ReorderProjectRequest(
                projectId: projectId,
                newStatus: status.rawValue,
                newOrder: order
            )
            let _: APIResponse<EmptyResponse> = try await apiClient.put(
                endpoint: .projectsReorder,
                body: request
            )

            // Update locally for immediate feedback
            if let index = projects.firstIndex(where: { $0.id == projectId }) {
                projects[index].status = status
                projects[index].order = order
            }
        } catch {
            await loadData() // Reload on error
        }
    }

    func deleteProject(id: String) async -> Bool {
        do {
            let response: APIResponse<EmptyResponse> = try await apiClient.delete(
                endpoint: .projects,
                queryParams: ["id": id]
            )
            if response.success {
                projects.removeAll { $0.id == id }
                return true
            }
        } catch {
            errorMessage = error.localizedDescription
        }
        return false
    }

    func addComment(projectId: String, text: String) async -> Bool {
        do {
            let request = AddCommentRequest(text: text)
            let response: APIResponse<EmptyResponse> = try await apiClient.post(
                endpoint: .projectComments(projectId),
                body: request
            )
            if response.success {
                await loadData()
                return true
            }
        } catch {
            errorMessage = error.localizedDescription
        }
        return false
    }
}
